File: HOT100/8.二叉树/236.二叉树的最近公共祖先.cpp
/*
 * @lc app=leetcode.cn id=236 lang=cpp
 *
 * [236] 二叉树的最近公共祖先
 */
#include "tools.h"
using namespace std;
struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        
        auto dfs = [&](auto&& self, TreeNode* n) -> TreeNode*
        {
            if (n == p || n == q || ! n) return n;
            TreeNode* l = self(self, n->left);
            TreeNode* r = self(self, n->right);
            if (l && r) return n;
            else if (l && ! r) return l;
            else if (! l && r) return r;
            else return NULL;
        };
        return dfs(dfs, root);
    }
};
// @lc code=end

File: 题单/&1/236.二叉树的最近公共祖先.py
#
# @lc app=leetcode.cn id=236 lang=python3
#
# [236] 二叉树的最近公共祖先
#
from mytools import *
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        def dfs(n, p, q):
            if not n: return None
            if n == p or n == q: return n
            l = dfs(n.left, p, q)
            r = dfs(n.right, p, q)
            if l and r: return n
            return l if l else r
        return dfs(root, p, q)
            
# @lc code=end

